picard/cluster.py

```diff
@@ -486,22 +521,33 @@ def get_cluster_title(self, cluster):
 
         return maxWord
 
-    def cluster(self, threshold):
-
+    def cluster(self, threshold, tagger=None):
         # Keep the matches sorted in a heap
         heap = []
+        num_files = self.cluster_dict.get_size()
+
+        # 20 evenly spaced indexes of files being clustered, used as checkpoints for every 5% progress
+        status_update_steps = set([int(i*num_files/20) for i in range(1, 21)])
 
-        for y in range(self.cluster_dict.get_size()):
+        for y in range(num_files):
             token_y = self.cluster_dict.get_token(y).lower()
             for x in range(y):
                 if x != y:
                     token_x = self.cluster_dict.get_token(x).lower()
                     c = similarity(token_x, token_y)
                     if c >= threshold:
                         heappush(heap, ((1.0 - c), [x, y]))
-            QtCore.QCoreApplication.processEvents()
-
-        for i in range(self.cluster_dict.get_size()):
+            if tagger and y in status_update_steps:
+                statusmsg = "Clustering - step %(step)d/3: %(cluster_type)s (%(update)d%%)"
+                mparams = {
+                    'step': self.cluster_type.value,
+                    'cluster_type': _(self._cluster_type_label()),
+                    'update': 100*(y+1)//num_files,
+                }
+                tagger.window.set_statusbar_message(statusmsg, mparams)
+                QtCore.QCoreApplication.processEvents()
+
+        for i in range(num_files):
```

Comment:
Cannot this loop merged with previous one?

Reply:
Yup. Just didn't want to change how the code was organized. I didn't even look into what those loops do.